processing/src/main/java/org/apache/druid/query/UnionDataSource.java

```diff
@@ -26,12 +26,19 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Iterables;
 import org.apache.druid.java.util.common.IAE;
+import org.apache.druid.timeline.Overshadowable;
+import org.apache.druid.timeline.TimelineLookup;
+import org.apache.druid.timeline.TimelineObjectHolder;
+import org.joda.time.Interval;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Map;
 import java.util.Set;
+import java.util.function.BiFunction;
 import java.util.stream.Collectors;
 
-public class UnionDataSource implements DataSource
+public class UnionDataSource implements MultiTableDataSource
```

Comment:
In the future, if we wanted to extend UnionDataSource to support unioning nontables (like, perhaps, unioning the results of queries) then do you see a good migration path to get there?

Reply:
In that case, I would expect us to have a `MultiTableDataSource` implementation say `UnionTableDataSource` and the current `UnionDataSource` would go back to being a `DataSource` implementation that would start handling a list of `DataSource` objects. 
Also, this would mean that we'd still have to keep the `UnionQueryRunner` to execute the nontable datasources and `UnionTableDataSource` separately and merge the results. Therefore for a union query with multiple tables and nontables, the tables would still be executed together under `UnionTableDataSource` and the non table datasources would be executed individually. There should be some refactoring within `ClientQuerySegmentWalker` in case we need to handle non tables that can be fulfilled with `LocalQuerySegmentWalker`.
Do you see any issues with the `MultiTableDataSource` design in this PR?